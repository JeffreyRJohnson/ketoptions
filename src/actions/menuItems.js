import uuid from 'uuid';
import database from '../firebase/firebase';

// ADD_MENU_ITEM
export const addMenuItem = menuItem => ({
  type: 'ADD_MENU_ITEM',
  menuItem
});

export const startAddMenuItem = (menuItemData = {}) => {
  return dispatch => {
    const {
      calories = 0,
      carbs = 0,
      fat = 0,
      menu_item = '',
      protein = 0,
      restaurant_name = ''
    } = menuItemData;

    const menuItem = {
      calories,
      carbs,
      fat,
      menu_item,
      protein,
      restaurant_name
    };

    return database
      .ref('menuItems')
      .push(menuItem)
      .then(ref => {
        dispatch(
          addMenuItem({
            id: ref.key,
            ...menuItem
          })
        );
      });
  };
};

// REMOVE_MENU_ITEM
export const removeMenuItem = ({ id } = {}) => ({
  type: 'REMOVE_MENU_ITEM',
  id
});

// EDIT_MENU_ITEM
export const editMenuItem = (id, updates) => ({
  type: 'EDIT_MENU_ITEM',
  id,
  updates
});
