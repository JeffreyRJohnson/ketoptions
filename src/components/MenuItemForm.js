import React from 'react';

export default class MenuItemForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      menu_item: props.menuItem ? props.menuItem.menu_item : '',
      restaurant_name: props.menuItem ? props.menuItem.restaurant_name : '',
      calories: props.menuItem ? props.menuItem.calories : '',
      protein: props.menuItem ? props.menuItem.protein : '',
      fat: props.menuItem ? props.menuItem.fat : '',
      carbs: props.menuItem ? props.menuItem.carbs : '',
      error: ''
    };
  }

  onRestaurantNameChange = e => {
    const restaurant_name = e.target.value;
    this.setState(() => ({ restaurant_name }));
  };

  onMenuItemChange = e => {
    const menu_item = e.target.value;
    this.setState(() => ({ menu_item }));
  };

  onCalorieChange = e => {
    const calories = e.target.value;
    this.setState(() => ({ calories }));
  };

  onProteinChange = e => {
    const protein = e.target.value;
    this.setState(() => ({ protein }));
  };

  onFatChange = e => {
    const fat = e.target.value;
    this.setState(() => ({ fat }));
  };

  onCarbChange = e => {
    const carbs = e.target.value;
    this.setState(() => ({ carbs }));
  };

  onSubmit = e => {
    e.preventDefault();

    if (
      !this.state.restaurant_name ||
      !this.state.menu_item ||
      !this.state.calories ||
      !this.state.protein ||
      !this.state.fat ||
      !this.state.carbs
    ) {
      this.setState(() => ({ error: 'You are missing a field' }));
    } else {
      this.setState(() => ({ error: '' }));
      this.props.onSubmit({
        restaurant_name: this.state.restaurant_name,
        menu_item: this.state.menu_item,
        calories: this.state.calories,
        protein: this.state.protein,
        fat: this.state.fat,
        carbs: this.state.carbs
      });
    }
  };

  render() {
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmit}>
          <input
            type="text"
            placeholder="Restaurant name"
            autoFocus
            value={this.state.restaurant_name}
            onChange={this.onRestaurantNameChange}
          />
          <input
            type="text"
            placeholder="Menu Item name"
            value={this.state.menu_item}
            onChange={this.onMenuItemChange}
          />
          <input
            type="number"
            placeholder="Calories"
            value={this.state.calories}
            onChange={this.onCalorieChange}
          />
          <input
            type="number"
            placeholder="Protein"
            value={this.state.protein}
            onChange={this.onProteinChange}
          />
          <input
            type="number"
            placeholder="Fat"
            value={this.state.fat}
            onChange={this.onFatChange}
          />
          <input
            type="number"
            placeholder="Carbs"
            value={this.state.carbs}
            onChange={this.onCarbChange}
          />
          <button>Add Menu Item</button>
        </form>
      </div>
    );
  }
}
