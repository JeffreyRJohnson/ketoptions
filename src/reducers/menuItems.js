// MenuItems Reducer

const menuItemsReducerDefaultState = [];

export default (state = menuItemsReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_MENU_ITEM':
      return [...state, action.menuItem];
    case 'REMOVE_MENU_ITEM':
      return state.filter(({ id }) => id !== action.id);
    case 'EDIT_MENU_ITEM':
      return state.map(menuItem => {
        if (menuItem.id === action.id) {
          return {
            ...menuItem,
            ...action.updates
          };
        } else {
          return menuItem;
        }
      });
    default:
      return state;
  }
};
