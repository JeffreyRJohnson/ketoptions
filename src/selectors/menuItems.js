// Get visible menuItems

export default (menuItems, { text }) => {
  return menuItems.filter(menuItem => {
    const textMatch = menuItem.restaurant_name
      .toLowerCase()
      .includes(text.toLowerCase());

    return textMatch;
  });
};
