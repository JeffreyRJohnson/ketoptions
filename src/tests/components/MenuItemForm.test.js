import React from 'react';
import { shallow } from 'enzyme';
import MenuItemForm from '../../components/MenuItemForm';
import menuItems from '../fixtures/menuItems';

test('should render MenuItemForm correctly', () => {
  const wrapper = shallow(<MenuItemForm />);
  expect(wrapper).toMatchSnapshot();
});

test('should render MenuItemForm correctly with menuItems', () => {
  const wrapper = shallow(<MenuItemForm menuItem={menuItems[1]} />);
  expect(wrapper).toMatchSnapshot();
});

test('should render error for invalid form submission', () => {
  const wrapper = shallow(<MenuItemForm />);
  wrapper.find('form').simulate('submit', {
    preventDefault: () => {}
  });
  expect(wrapper.state('error').length).toBeGreaterThan(0);
  expect(wrapper).toMatchSnapshot();
});

test('should set restaurant_name on input change', () => {
  const value = 'New restaurant name';
  const wrapper = shallow(<MenuItemForm />);
  wrapper
    .find('input')
    .at(0)
    .simulate('change', {
      target: { value }
    });
  expect(wrapper.state('restaurant_name')).toBe(value);
});

test('should set menu_item on input change', () => {
  const value = 'New menu item';
  const wrapper = shallow(<MenuItemForm />);
  wrapper
    .find('input')
    .at(1)
    .simulate('change', {
      target: { value }
    });
  expect(wrapper.state('menu_item')).toBe(value);
});

test('should set calories on input change', () => {
  const value = 5;
  const wrapper = shallow(<MenuItemForm />);
  wrapper
    .find('input')
    .at(2)
    .simulate('change', {
      target: { value }
    });
  expect(wrapper.state('calories')).toBe(value);
});

test('should set protein on input change', () => {
  const value = 5;
  const wrapper = shallow(<MenuItemForm />);
  wrapper
    .find('input')
    .at(3)
    .simulate('change', {
      target: { value }
    });
  expect(wrapper.state('protein')).toBe(value);
});

test('should set fat on input change', () => {
  const value = 5;
  const wrapper = shallow(<MenuItemForm />);
  wrapper
    .find('input')
    .at(4)
    .simulate('change', {
      target: { value }
    });
  expect(wrapper.state('fat')).toBe(value);
});

test('should set carbs on input change', () => {
  const value = 5;
  const wrapper = shallow(<MenuItemForm />);
  wrapper
    .find('input')
    .at(5)
    .simulate('change', {
      target: { value }
    });
  expect(wrapper.state('carbs')).toBe(value);
});

test('should call onSubmit prop for valid form submission', () => {
  const onSubmitSpy = jest.fn();
  const wrapper = shallow(
    <MenuItemForm menuItem={menuItems[0]} onSubmit={onSubmitSpy} />
  );
  wrapper.find('form').simulate('submit', {
    preventDefault: () => {}
  });
  expect(wrapper.state('error')).toBe('');
  expect(onSubmitSpy).toHaveBeenLastCalledWith({
    restaurant_name: menuItems[0].restaurant_name,
    menu_item: menuItems[0].menu_item,
    calories: menuItems[0].calories,
    protein: menuItems[0].protein,
    fat: menuItems[0].fat,
    carbs: menuItems[0].carbs
  });
});
