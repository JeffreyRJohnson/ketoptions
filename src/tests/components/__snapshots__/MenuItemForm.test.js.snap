// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render MenuItemForm correctly 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <input
      autoFocus={true}
      onChange={[Function]}
      placeholder="Restaurant name"
      type="text"
      value=""
    />
    <input
      onChange={[Function]}
      placeholder="Menu Item name"
      type="text"
      value=""
    />
    <input
      onChange={[Function]}
      placeholder="Calories"
      type="number"
      value=""
    />
    <input
      onChange={[Function]}
      placeholder="Protein"
      type="number"
      value=""
    />
    <input
      onChange={[Function]}
      placeholder="Fat"
      type="number"
      value=""
    />
    <input
      onChange={[Function]}
      placeholder="Carbs"
      type="number"
      value=""
    />
    <button>
      Add Menu Item
    </button>
  </form>
</div>
`;

exports[`should render MenuItemForm correctly with menuItems 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <input
      autoFocus={true}
      onChange={[Function]}
      placeholder="Restaurant name"
      type="text"
      value="Wendys"
    />
    <input
      onChange={[Function]}
      placeholder="Menu Item name"
      type="text"
      value="chilli - small"
    />
    <input
      onChange={[Function]}
      placeholder="Calories"
      type="number"
      value={300}
    />
    <input
      onChange={[Function]}
      placeholder="Protein"
      type="number"
      value={30}
    />
    <input
      onChange={[Function]}
      placeholder="Fat"
      type="number"
      value={15}
    />
    <input
      onChange={[Function]}
      placeholder="Carbs"
      type="number"
      value={3}
    />
    <button>
      Add Menu Item
    </button>
  </form>
</div>
`;

exports[`should render error for invalid form submission 1`] = `
<div>
  <p>
    You are missing a field
  </p>
  <form
    onSubmit={[Function]}
  >
    <input
      autoFocus={true}
      onChange={[Function]}
      placeholder="Restaurant name"
      type="text"
      value=""
    />
    <input
      onChange={[Function]}
      placeholder="Menu Item name"
      type="text"
      value=""
    />
    <input
      onChange={[Function]}
      placeholder="Calories"
      type="number"
      value=""
    />
    <input
      onChange={[Function]}
      placeholder="Protein"
      type="number"
      value=""
    />
    <input
      onChange={[Function]}
      placeholder="Fat"
      type="number"
      value=""
    />
    <input
      onChange={[Function]}
      placeholder="Carbs"
      type="number"
      value=""
    />
    <button>
      Add Menu Item
    </button>
  </form>
</div>
`;
